config/session.go

```diff
@@ -112,3 +128,292 @@ func (store *SessionStore) LoadSessionStateAndCheckIDP(r *http.Request) (*sessio
 func (store *SessionStore) SaveSession(w http.ResponseWriter, r *http.Request, v any) error {
 	return store.store.SaveSession(w, r, v)
 }
+
+var (
+	accessTokenUUIDNamespace   = uuid.MustParse("0194f6f8-e760-76a0-8917-e28ac927a34d")
+	identityTokenUUIDNamespace = uuid.MustParse("0194f6f9-aec0-704e-bb4a-51054f17ad17")
+)
+
+type IncomingIDPTokenSessionCreator interface {
+	CreateSession(ctx context.Context, cfg *Config, policy *Policy, r *http.Request) (*session.Session, error)
+}
+
+type incomingIDPTokenSessionCreator struct {
+	getRecord  func(ctx context.Context, recordType, recordID string) (*databroker.Record, error)
+	putRecords func(ctx context.Context, records []*databroker.Record) error
+}
+
+func NewIncomingIDPTokenSessionCreator(
+	getRecord func(ctx context.Context, recordType, recordID string) (*databroker.Record, error),
+	putRecords func(ctx context.Context, records []*databroker.Record) error,
+) IncomingIDPTokenSessionCreator {
+	return &incomingIDPTokenSessionCreator{getRecord: getRecord, putRecords: putRecords}
+}
+
+// CreateSession attempts to create a session for incoming idp access and
+// identity tokens. If no access or identity token is passed ErrNoSessionFound will be returned.
+// If the tokens are not valid an error will be returned.
+func (c *incomingIDPTokenSessionCreator) CreateSession(
+	ctx context.Context,
+	cfg *Config,
+	policy *Policy,
+	r *http.Request,
+) (session *session.Session, err error) {
+	if rawAccessToken, ok := cfg.GetIncomingIDPAccessTokenForPolicy(policy, r); ok {
+		return c.createSessionAccessToken(ctx, cfg, policy, rawAccessToken)
+	}
+
+	if rawIdentityToken, ok := cfg.GetIncomingIDPIdentityTokenForPolicy(policy, r); ok {
+		return c.createSessionForIdentityToken(ctx, cfg, policy, rawIdentityToken)
+	}
+
+	return nil, sessions.ErrNoSessionFound
+}
+
+func (c *incomingIDPTokenSessionCreator) createSessionAccessToken(
+	ctx context.Context,
+	cfg *Config,
+	policy *Policy,
+	rawAccessToken string,
+) (*session.Session, error) {
+	sessionID := uuid.NewSHA1(accessTokenUUIDNamespace, []byte(rawAccessToken)).String()
+	s, err := c.getSession(ctx, sessionID)
```

Comment:
should the session id be tied the route as well? what happens if you use the same access token for different routes?

Reply:
Sessions are typically shared across routes, but I could make it route-specific.

Reply:
If you use a token that authenticates you to one route, then try to use it for a different route (with a separate policy) it would still check that route's policy, right? if multiple routes have identical policy though then i suppose it makes sense to share the session

Reply:
These sessions will now be route-specific.

If we don't think sessions should be shared across routes in Pomerium we should create a separate follow up issue. I thought session sharing was a feature of the product.

Reply:
If it makes more sense to have the sessions shared, then they should be shared. I'm not sure what would be best. If the session persists, is it just that the user's identity is cached? or does it affect whether the user is authorized for different routes?

Reply:
I don't know what would be best either. I guess I can update it so the session is only shared based on the identity provider config since that can vary by routes (but doesn't usually).

A session is used to track authentication. It tells us who the user is. It does not grant access to routes unless the "any authenticated user" policy is used.